### PR TITLE
Implement undo & redo stacks in Graphical Recording Mode

### PR DESCRIPTION
**Pull request by gestureleft**
Certain actions can now be undone and redone. The actions are maintained on a "stacks" that the UI logic can push to and pop from

Written by Harry + Dinu

### CODE REVIEW

File: packages/TorneloScoresheet/src/util/moveHistory.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+  // Last ply = MovePly
+  return (
+    chessEngine.makeMove(
+      lastPly.startingFen,
+      lastPly.move,
+      lastPly.promotion,
+    ) ?? '' // all move in history are legal, -> should never return undef
+  );
+};
```
Comment (nocommentcode) [inline packages/TorneloScoresheet/src/util/moveHistory.ts:31]: Should you remove this function from the `recordingState` file then and use the util version from the recording hook?

**Review by nocommentcode — CHANGES_REQUESTED** [review]
There are comipilation errors in enterPgn page, where you changed the state but not the page,
Its crashing when trying to undo, getting a 'stack.at is undefined' error
Also it is significantly slower now, it takes like 2 of 3 seconds to process a move i dont think this is acceptable
Also the are we not going to support undoing editing moves?
Also there are no unit tests for the undo stack hook

File: packages/TorneloScoresheet/src/util/moveHistory.ts (hunk 1 above)
Comment (gestureleft) [reply packages/TorneloScoresheet/src/util/moveHistory.ts]: This has been removed from the `recordingState` file

**Comment by gestureleft** [comment]
> There are comipilation errors in enterPgn page, where you changed the state but not the page,

Compilation errors fixed.

> Its crashing when trying to undo, getting a 'stack.at is undefined' error

This is strange - `stack.at` is definitely [a thing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at) and is working on my machine. I guess it's possible that the Android version of Hermes doesn't support it yet for some reason?

> Also it is significantly slower now, it takes like 2 of 3 seconds to process a move i dont think this is acceptable

The performance is identical on my machine. I don't see why there would be any change - the only extra overhead is pushing to an array.

> Also the are we not going to support undoing editing moves?

Yep - this PR is already big enough though, so that'll be added in a future PR :^)

> Also there are no unit tests for the undo stack hook

Good point - I'll get on this now!

**Review by nocommentcode — APPROVED** [review]
Its working for me now,
Its still considerable slower now though...

**Pull request merged**